pennylane/transforms/split_non_commuting.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the tape transform that splits non-commuting terms
+"""
+# pylint: disable=protected-access
+import pennylane as qml
+import numpy as np
+
+
+def split_non_commuting(tape):
+    r"""
+    Splits a tape measuring non-commuting observables into groups of commuting observables.
+
+    Args:
+        tape (.QuantumTape): the tape used when calculating the expectation value
+            of the Hamiltonian
+
+    Returns:
+        tuple[list[.QuantumTape], function]: Returns a tuple containing a list of
+        quantum tapes to be evaluated, and a function to be applied to these
+        tape executions to restore the ordering of the inputs.
+
+    **Example**
+
+    We can create a tape with non-commuting observables:
+
+    .. code-block:: python3
+
+        import pennylane as qml
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliY(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    Now ``tapes`` is a list of two tapes, each for one of the non-commuting terms:
+
+    >>> [t.observables for t in tapes]
+    [[expval(PauliZ(wires=[0]))], [expval(PauliY(wires=[0]))]]
+
+    The processing function becomes important when creating the commuting groups as the order of the inputs has been modified:
+
+    .. code-block:: python3
+
+        with qml.tape.QuantumTape() as tape:
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+            qml.expval(qml.PauliX(0) @ qml.PauliX(1))
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliX(0))
+        tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+    In this example, the groupings are ``group_coeffs = [[0,2], [1,3]]`` and processing_fn makes sure that the final output is of the same shape and ordering:
+
+    >>> processing_fn(tapes)
+    tensor([tensor(expval(PauliZ(wires=[0]) @ PauliZ(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0]) @ PauliX(wires=[1])), dtype=object, requires_grad=True),
+        tensor(expval(PauliZ(wires=[0])), dtype=object, requires_grad=True),
+        tensor(expval(PauliX(wires=[0])), dtype=object, requires_grad=True)],
+       dtype=object, requires_grad=True)
```

Comment:
@Qottmann if you take my suggestion above about making this a `batch_transform`, then it might make more sense to make the main example use QNodes, not tapes, since these are more familiar to users. For example:

````rest
This transform allows us to transform a QNode that measures
non-commuting observables to *multiple* circuit executions
with qubit-wise commuting groups:

.. code-block:: python3

    @qml.transforms.split_non_commuting
    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(0))

>>> circuit(0.2)
tensor([ 0.98006658, -0.19866933], requires_grad=True)
>>> print(qml.draw(circuit)(0.2))
0: ──RX(0.20)─┤  <Z>
0: ──RX(0.20)─┤  <Y>
````

You could keep the above tape example, but move it into a 'usage details' section, below this example, e.g.,

```rest
.. details::
    :title: Usage Details

    This transform also works with tapes. We can create a tape with non-commuting observables:
```
etc.

Reply:
See https://pennylane.readthedocs.io/en/latest/code/api/pennylane.adjoint.html for another example of a function that uses `Usage Details` :)